src/replication.c

```diff
@@ -1914,6 +1938,204 @@ disconnect_nonresponding_replica(replica_t *replica, uint64_t io_seq,
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica->zvol_guid == guid) {
+			return replica->replica_id;
+		}
+	}
+	return NULL;
+}
+
+static void
+process_snaplist_response(spec_t *spec, char **resp, void *buf, int rcount)
+{
+	struct zvol_snapshot_list **snap_resp_array = (struct zvol_snapshot_list **)buf;
+	int i = 0;
+	uint64_t total_len;
+	char *msg = NULL, *replica_id = NULL;
+	const char *json_string = NULL;
+	struct json_object *jobj, *jarray, *resp_obj, *snap_obj;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	jarray = json_object_new_array();
+
+	for (i = 0; i < rcount; i += 1) {
+		if (snap_resp_array[i]) {
+			resp_obj = snap_obj = NULL;
+			jobj = json_object_new_object();
+
+			replica_id = get_replica_id_for_guid(spec, snap_resp_array[i]->zvol_guid);
+			if(!replica_id) {
+				REPLICA_ERRLOG("No replica_id for replica:%lu", snap_resp_array[i]->zvol_guid);
+				continue;
+			}
+
+			json_object_object_add(jobj, "replica_id", json_object_new_string(replica_id));
+
+			resp_obj = json_tokener_parse(snap_resp_array[i]->data);
+			(void) json_object_object_get_ex(resp_obj,
+			    "snapshot", &snap_obj);
+			json_object_object_add(jobj, "snapshot", snap_obj);
+
+			json_object_array_add(jarray, jobj);
+		}
+	}
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "snapshots", jarray);
+
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+	total_len = strlen(json_string) + 1;
+	msg = malloc(total_len);
+	memset(msg, 0, total_len);
+	strncpy(msg, json_string, total_len);
+	json_object_put(jobj);
+	*resp = msg;
+}
+
+char *
+istgt_lu_fetch_snaplist(spec_t *spec, int wait_time, char *snapname)
+{
+	replica_t *replica;
+	int free_rcomm_mgmt = 0;
+	struct timespec last, now, diff;
+	rcommon_mgmt_cmd_t *rcomm_mgmt;
+	mgmt_cmd_t *mgmt_cmd;
+	size_t data_len;
+	char *response = NULL;
+	int num_replica = 0;
+	zvol_io_hdr_t *mgmt_io_hdr = NULL;
+	uint64_t num;
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+
+	MTX_LOCK(&spec->rq_mtx);
+	num_replica = spec->healthy_rcount + spec->degraded_rcount;
+	rcomm_mgmt = allocate_rcommon_mgmt_cmd(sizeof (struct zvol_snapshot_list *) * num_replica);
+	free_rcomm_mgmt = 0;
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+		memset(mgmt_cmd, 0, sizeof (mgmt_cmd_t));
+
+		data_len = strlen(spec->volname) + ((snapname) ? strlen(snapname) + 1: 0) + 2;
+
+		BUILD_REPLICA_MGMT_HDR(mgmt_io_hdr, ZVOL_OPCODE_SNAP_LIST, data_len);
+
+		mgmt_cmd->data = (char *)malloc(data_len);
+		if (snapname)
+			snprintf(mgmt_cmd->data, data_len, "%s@%s", spec->volname, snapname);
+		else
+			snprintf(mgmt_cmd->data, data_len, "%s", spec->volname);
+
+		mgmt_cmd->rcomm_mgmt = rcomm_mgmt;
+		mgmt_cmd->io_hdr = mgmt_io_hdr;
+		mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+		MTX_LOCK(&replica->r_mtx);
+		TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+		MTX_UNLOCK(&replica->r_mtx);
+
+		rcomm_mgmt->cmds_sent++;
```

Comment:
how about incrementing this before adding to mgmt_cmd_queue?

Reply:
We can do that, but i think doing increment after the command gets queued will be good, logically. WDYT?